### PR TITLE
Remove pro role + expose analytics to all users via dashboard

### PR DESCRIPTION
**Pull request by vaidehijoshi**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [x] Documentation Update

## Related Tickets & Documents

Closes https://github.com/forem/internalEngineering/issues/403.

## QA Instructions, Screenshots, Recordings

To QA this feature, first make sure that you have a user locally who _does not_ have the `:pro` role:
```ruby
pry(main)> u = User.find_by(username: 'your-username-goes-here')
=> #<User id: ...>
]pry(main)> u.has_role? :pro
=> false
```

**QA Instructions, Part I:**
1. Try going to `/dashboard`. On the bottom left of the sidebar, you should see an `Analytics` option in the menu. Click on that:
Screen_Shot_2021-03-28_at_4_00_05_PM

2. Clicking on that should take you to your "Analytics Dashboard", which previously was the "Pro dashboard". Make sure there is no "pro" language (everything should read "analytics dashboard):
Screen Shot 2021-03-28 at 4 00 24 PM
3. Now, go to any article you have published (if you haven't published any, create a quick test post). 
Screen_Shot_2021-03-28_at_4_00_50_PM
4. You should see a dashboard to manage your post. Notably, you should see a `Stats` button (which used to only be rendered for "pro" users). Click on that button:
Screen_Shot_2021-03-28_at_4_01_13_PM
5. You'll now be able to see the stats for your article on a `/stats` route, which should render analytics that are _specific_ to the article you were just looking at:
Screen Shot 2021-03-28 at 4 01 24 PM

**QA Instructions, Part II:**
1. Create an organization ([here's the route to do so](http://localhost:3000/settings/organization/new)).
2. Go back to your `/dashboard`. Note that the `Analytics` option still renders as before, only now you _also_ should see an `Analytics [Your Org Name` option as well (click on that one!):
Screen_Shot_2021-03-28_at_4_04_42_PM
3. Now, you should see the analytics dashboard for _your_ specific organization:
Screen Shot 2021-03-28 at 4 31 19 PM
4. Write a post under your organization and go to that post, and click `Manage` to go to the dashboard for your newly-published article, which was published under your specific organization (the route should look something like `/llamafans/this-post-is-for-llama-fans-only-43fl/manage`. You should see the organization name render in the top left corner, and you should see a `Stats` button as well. Click on the `Stats` button:
Screen_Shot_2021-03-28_at_4_04_53_PM
5. You should be taken to an article-specific analytics dashboard for the organization you are a part of!
Screen Shot 2021-03-28 at 4 05 02 PM

**QA Instructions, Part III:**
1. Assign your user either an `:admin` or `:super_admin` role, and go to `/admin/users`. Select a user to edit, and scroll down to the "User Status" section of the page. Click the dropdown, and ensure that the `Pro` role _does not_ show up there:
Screen Shot 2021-03-28 at 3 58 48 PM
2. Feel free to check for anywhere else on the site or in the codebase where the notion of a `Pro` role still exists (hopefully I got all of them, but please let me know if I missed anything! 😃)

### UI accessibility concerns?

_No HTML or UX changes happens as a part of this PR, so I don't think so._

## Added tests?

- [ ] Yes
- [x] No, and this is why: _I've updated pre-existing tests and added new ones where necessary (but didn't overhaul the testing significantly enough IMO to check the "Yes" box 😉)_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [x] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post (**I will work with @cmgorton to make sure this is highlighted/announced to the community since we are sharing a feature that was previously only available to handful of folks!**)
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## Are there any post deployment tasks we need to perform?

After this feature is out, we will need to delete any `Pro` roles from the database. Even with the "pro" code all removed, the role still exists in the database, as evidenced by this screenshot of my local Forem instance, once I added the `:pro` role to my own user and deleted the "pro" code (seems like the `Role` instance still exists in the database):
Screen Shot 2021-03-28 at 3 59 01 PM

 I _think_ that this only applies to the DEV community, but I'm not entirely sure about that and could use some gut checks there. My thought was to either write a DataUpdateScript (if this applies to many communities), otherwise maybe we can delete the `:pro` role directly via a DEV console?

A script would definitely be more comprehensive, but if no other Forem instance has `pro` roles, that feels much more heavyweight than we need it to be.

## What gif best describes this PR or how it makes you feel?

animation of woman cutting bangs off with scissors

### CODE REVIEW

File: app/policies/organization_policy.rb
Comment (vaidehijoshi) [inline app/policies/organization_policy.rb:36]: This was [only ever used](https://github.com/forem/forem/pull/13156/files#diff-8176f7cf6d68ad6630418f9d9faf7be510e5fb6e5e0bb535fe36efe3ed36bcb5L61) by the `DashboardsController#pro` action, so we can remove this now! ✂️ ✅

File: app/views/dashboards/analytics.erb
Comment (vaidehijoshi) [inline app/views/dashboards/analytics.erb:1]: I chose to keep this dashboard template/view the exact same and just renamed it from "Pro Analytics" to "Analytics". It seemed a little weird to extract it out into the basic `/dashboard` view, so keeping it as a separate route `/dashboard/analytics` seemed like a smoother transition, IMO.

File: config/elasticsearch/mappings/feed_content.json
Comment (vaidehijoshi) [inline config/elasticsearch/mappings/feed_content.json:175]: Our ES docs made it sound like it is safe/acceptable to edit these mapping files directly, so I went ahead and removed the `pro` key from them. Let me know if I misunderstood and actually shouldn't be touching these files for some reason!

File: config/elasticsearch/mappings/reactions.json
Comment (vaidehijoshi) [inline config/elasticsearch/mappings/reactions.json:83]: Our ES docs made it sound like it is safe/acceptable to edit these mapping files directly, so I went ahead and removed the `pro` key from them. Let me know if I misunderstood and actually shouldn't be touching these files for some reason!

File: docs/backend/roles.md
Comment (vaidehijoshi) [inline docs/backend/roles.md:36]: This just updates the docs so that the example we use for "assigning user roles" isn't an example that wouldn't actually work in the console for someone locally! 😛

File: spec/requests/articles/articles_spec.rb
Comment (vaidehijoshi) [inline spec/requests/articles/articles_spec.rb:307]: This test no longer applies since we are no longer authorizing against the `:pro` role at all for the `/stats` route of an article, so I removed it!

**Comment by vaidehijoshi** [comment]
Interesting, for some reason an E2E test [is failing](https://travis-ci.com/github/forem/forem/jobs/494360114#L2227) in what appears to be a completely unrelated test. I'm seeing a test failure that looks like it is [ElasticSearch related](https://travis-ci.com/github/forem/forem/jobs/494360114#L2228); the test seems to look at `search/usernames.json`, which I haven't touched, so I'm not quite sure at first glance why it is failing in this way. 🤔 

I am still going to mark it ready for review as I still think the rest of the feature can be reviewed, and maybe someone will have some ideas as to why this spec is failing. Otherwise, I can reach out to @nickytonline tomorrow to see if he has seen this failure before or knows of any good first steps to begin debugging it :)

**Comment by citizen428** [comment]
I did a quick check in the DB and we have 101 pro users remaining:

```sql
select count(*) from users_roles where role_id = (select id from roles where name = 'pro');
-- 101
```

This PR should probably include a data update script to remove them, or were you planning on doing that later?

**Review by rhymes — COMMENTED** [review]
A few things:

- I forgot that by removing the pro role we would expose analytics to everybody via the API as well, not just the UI. Though some calculations in `AnalyticsService` are cached, we might want to alert SRE as a post deploy task, so that they are alerted there might be a spike in usage when we announce.

- I found a bunch of other instances were the word `pro` appears (you can find them with `rg -i pro -w` in the console): `pro-container` and `pro-header` as CSS classes used in a couple of files, `app/javascript/readingList/__tests__/readingList.test.jsx`, `app/controllers/admin/users_controller.rb` has `pro` in the `user_params`, `app/services/data_sync/elasticsearch/user.rb` has knowledge of `pro`, `pro-checkmark` CSS class, `app/assets/javascripts/initializers/initializeBaseUserData.js` checks for `user.pro` to add the stats button.

I'll modify `app/serializers/search/reading_list_article_serializer.rb`'s comment after this is merged as it relies on assumptions based on `master`

File: config/elasticsearch/mappings/feed_content.json
Comment (mstruve) [reply config/elasticsearch/mappings/feed_content.json:175]: No reindexing needed. The field will still exist in the index and on the documents until they are reindexed naturally but it won't hurt anything. If we had billions of documents and were keeping ES then it would be worth it to explicitly remove it to save space but in our case, no additional action is needed here.

**Comment by vaidehijoshi** [comment]
> This PR should probably include a data update script to remove them, or were you planning on doing that later?

@citizen428 Heh, I had thought of this and included a note about it in the PR description at the bottom (`Are there any post deployment tasks we need to perform?`):

> After this feature is out, we will need to delete any Pro roles from the database. Even with the "pro" code all removed, the role still exists in the database, as evidenced by this screenshot of my local Forem instance, once I added the :pro role to my own user and deleted the "pro" code (seems like the Role instance still exists in the database):
> Screen Shot 2021-03-28 at 3 59 01 PM
> 
> I think that this only applies to the DEV community, but I'm not entirely sure about that and could use some gut checks there. My thought was to either write a DataUpdateScript (if this applies to many communities), otherwise maybe we can delete the :pro role directly via a DEV console?
> 
> A script would definitely be more comprehensive, but if no other Forem instance has pro roles, that feels much more heavyweight than we need it to be.

I hadn't yet checked how many users had the pro role (for some reason I thought it was only like, 5 users?), so thank for checking on that for me! 🙏🏽 

Given that information, I think I would prefer just writing a DataUpdateScript for this rather than trying to do anything in the console. I have a draft PR ready, which will remove the `pro` role from any users who have it currently, and subsequently will delete the role entirely from the database: https://github.com/forem/forem/pull/13176

I think it makes sense to deploy _this PR_ first and make sure that it is working as expected before we remove the role from users + delete it from the database.

**Comment by vaidehijoshi** [comment]
> we might want to alert SRE as a post deploy task, so that they are alerted there might be a spike in usage when we announce.

Oh, great call! I'll be sure to loop in @forem/sre when we write the forem.dev post and announce this so they'll be aware of any spike in traffic

> * I found a bunch of other instances were the word `pro` appears (you can find them with `rg -i pro -w` in the console)

Omg, thank you! I thought I had searched all the variations of `pro` but clearly missed some. ALSO TIL about `ripgrep`, that is _so handy_.

> I'll modify `app/serializers/search/reading_list_article_serializer.rb`'s comment after this is merged as it relies on assumptions based on `master`

Sounds good 👍🏽 

@rhymes I have updated this to remove all other instances of `pro`, let me know if I missed anything. I also have a data update script ready to go here: https://github.com/forem/forem/pull/13176, but I think we should wait to merge that until this one is out, the forem.dev post is live (working on that today), and until we've deployed this out to the fleet and know that it works the way it should. I'm always cautious to _not_ remove things from the DB until later, when we know it's safe to do so 😅

**Comment by citizen428** [comment]
@vaidehijoshi 

> Heh, I had thought of this and included a note about it in the PR description at the bottom 

Sorry, I completely missed this.

File: app/views/dashboards/_dashboard_article.html.erb
Comment (citizen428) [inline app/views/dashboards/_dashboard_article.html.erb:68]: I wonder if we should add this to our URL helper module, e.g. 

```ruby
# app/lib/url.rb
def self.article_stats(article)
  URI.join(url(article.path), "stats")
end

# here
<a href="<%= URL.article_stats(article) %>" data-no-instant class="crayons-btn">Stats</a>
```

It's not super important, but it would be a bit more in line with how we deal with similar URLs elsewhere. This can of course be done in a separate PR.

**Review by rhymes — APPROVED** [review]
Let's do this!

File: app/views/dashboards/_dashboard_article.html.erb
Comment (vaidehijoshi) [reply app/views/dashboards/_dashboard_article.html.erb:68]: I might do this as a 15 minute follow up fix! I'm having really frustrating issues with Travis today and I kind of want to hold off on triggering another build here for fear of Travis just not picking it up. Plus, as you said, this could be a great follow up PR 👍🏽

**Comment by vaidehijoshi** [comment]
This PR needs one more review 😎 Pinging @jgaskins for review since he has already reviewed the follow up to _this_ PR, https://github.com/forem/forem/pull/13176.

**Review by jgaskins — APPROVED** [review]
Oh I think I got confused before, seeing 2 PRs with similar diffs, and assumed the other one superseded this one. :joy: This looks good! :100:

**Comment by vaidehijoshi** [comment]
> Oh I think I got confused before, seeing 2 PRs with similar diffs, and assumed the other one superseded this one. 😂 This looks good! 💯

@jgaskins That was totally my fault! I forgot to set #13176's base branch when I opened the PR up initially 🙈 If you look at [that PR now](https://github.com/forem/forem/pull/13176#issuecomment-809752660), hopefully the changes make more sense. For full context, _this_ PR will be deployed first, and then once I've confirmed that it behaves the way that it should, I'll go ahead and merge #13176 to actually drop the old roles :)

**Pull request merged**
